### PR TITLE
[L1SpillManagement] gtest infra + 15 tests for the L1 spill pass

### LINKED ISSUE #8514: [L1SpillMgmt] Add unit test infra for L1 spill management pass
## Summary

The L1 spill management pass (`L1SpillManagement.cpp`) contains several interacting mechanisms — memory tracker, event-log snapshots, Belady eviction, reshard insertion, spill-to-DRAM, and the revalidation cascade — but there is currently no way to unit-test these in isolation.

All existing coverage comes through full model compilation lit tests, which makes it hard to:
- Reproduce a specific failure case with a minimal graph
- Assert on a particular sequence of evictions or reshard insertions
- Verify snapshot/replay correctness after a targeted eviction
- Test the memory tracker's free-list and address aliasing logic directly

## Desired infra

Create a C++ unit-test fixture (gtest or llvm-lit based) that allows a test to:

1. **Build a custom op graph** with an arbitrary DAG of `ttnn` ops (e.g. `MatmulOp`, `AddOp`, `ToLayoutOp`, `ReshardOp`) using the MLIR C++ API without going through the full optimizer pipeline.
2. **Assign explicit layouts** to every op result — both L1-sharded and DRAM — so a test fully controls the memory footprint of each live tensor.
3. **Run the pass directly** on the constructed module, supplying a mock system descriptor with configurable L1 budget per core.
4. **Inspect pass output** and assert on:
   - Which tensors were spilled to DRAM (`ToMemoryConfigOp` insertion)
   - Which reshards were inserted and at what schedule positions
   - Final `ScheduleData` contents (positionMap, deathSchedule)
   - `SumL1MemoryTracker` state (occupied bytes, free-list)
   - Snapshot/replay log entries via the `L1SpillObserver` interface

## Why this matters

The pass interacts with OpModel validation, address replay, and MLIR IR mutation in ways that are very hard to exercise via lit tests alone.  A targeted graph-builder fixture would let us:
- Reproduce the schedule-data staleness bug (see companion issue) in a two-line test
- Catch regressions in Belady eviction ordering
- Validate CB-fragmentation cushion behaviour with exact L1 budgets

## Suggested approach

- Add a `test/unittests/Dialect/TTNN/L1SpillManagementTest.cpp` file.
- Provide a `L1SpillTestBuilder` helper that wraps `mlir::OpBuilder` and exposes `addMatmul()`, `addEltwiseBinary()`, `setLayout()`, `setL1Budget()`, etc.
- Mock the `OpConstraintValidator` or inject a fake that returns fixed `ValidationResult` values (size, cbPeakUsage) so tests are deterministic without a real device.
- Implement a `RecordingObserver` (implements `L1SpillObserver`) that captures all observer callbacks into a list for assertions.

## Acceptance criteria

- [ ] `L1SpillTestBuilder` can construct a linear chain: `A → B → C` with configurable tensor sizes and L1 budget
- [ ] Can force an OOM at op B and assert that A is evicted (Belady, A is the farthest-use tensor)
- [ ] Can assert a `ToMemoryConfigOp` (reshard) was inserted between B and C
- [ ] Snapshot/replay produces consistent memory tracker state (occupied == sum of live L1 tensors)
- [ ] Tests run under `cmake --build build --target check-ttmlir`

### PR DESCRIPTION
**Pull request by rpavlovicTT**
## Ticket

Fixes #8514 

## Summary

Adds a gtest fixture and 15 unit tests for `L1SpillManagement`, plus two small production-side changes that make the infra possible.

**Production (2 commits):**
- `feat(l1spill)`: add `SumL1MemoryTracker::backendValidator` (`std::function` hook). All three op_model call sites in the pass now route through `validateBackendDirect()`, so a single test-time hook covers `validate()`, the consumer-reshard probe in `evictValue`, and the DRAM CB re-query in `evictForDramCBGrowth`. Default `backendValidator == nullptr` preserves current behaviour. Exposes `getMemoryTracker()` accessor.
- `refactor(l1spill)`: split `spillToDram(Value, Operation *insertBefore = nullptr)` into `spillToDram(Value)` (snapshot/replay-safe) and `spillToDramBeforeTrigger(Value, Operation *)` (requires a full tracker reset; today only used by `evictAllFromL1`). Adds an `assert(memoryTracker.getOccupiedL1() == 0)` after the reset to enforce the invariant.

**Test infra:**
- `test/unittests/Optimizer/L1SpillTestFixture.h` — `RecordingObserver` capturing every `L1SpillObserver` callback + `L1SpillTestFixture` with per-op config (`setL1Usage`, `setDramCBPeak`, `forceOOM`, `forceNotImplemented`), graph builders (`beginFunc`, `addUnary`, `addBinary`, `addReshape`, `finishFunc`), `run()` that installs the backend lambda, and IR-inspection helpers.
- `kKiB` constant (`1024`); sizes read as e.g. `1300 * kKiB`.

**15 tests, organised by what they cover:**

| Path | Test |
|------|------|
| OOM → single-victim eviction | LinearChainTest |
| Farthest-last-use pick | FarthestLastUseOrderingTest |
| Self-spill (output > budget) | SelfSpillTest |
| evictAllFromL1 (NotImplemented) | NotImplementedTest |
| Post-flush tracker reset invariant | NotImplementedPostFlushTest |
| markEvictedAndRebuild + replay forward | SnapshotReplayManyAllocsTest |
| Snapshot refresh during replay | SnapshotReplayCrossEvictionTest |
| Non-empty snapshot restore + cascade replay past skipped events | SnapshotReplayNonEmptyAndCascadeTest |
| Reshard / past-consumer handling on fork | ForkJoinTest |
| No-spill invariants on realistic graphs | SequentialMLPTest, ResidualBlockTest, QKVForkJoinTest |
| handleFragmentation via cbPeakUsage | CBOverlapTest |
| Cushion-dominant fragmentation | CushionDominantTriggerTest |
| Free-list coalesce | HandleNoFitTest |

**1 DISABLED:** `ViewEligibleReshapeAliasTest` — reproducer for #8625 (alias path runs after the fit check; pass evicts a reshape's input at tight budgets instead of aliasing).

## Findings surfaced by this work

1. **#8625** — view-eligible reshape evicted instead of aliased at tight budgets.
2. **CushionDominantTriggerTest comment** — `ensureFitsL1` guards `wouldCBsOverlapTensors` behind `cbPeakUsage > 0` (line 597), so the function's "cushion alone with cbPeak=0" branch is unreachable from that caller. Documented in the test.

## Test plan

- [x] `./build/bin/L1SpillManagementTests` — 15 pass, 1 DISABLED.
- [x] `cmake --build build --target check-ttmlir` — 1775 passed, 0 failures.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: test/unittests/Optimizer/L1SpillTestFixture.h [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+      // tt-metal (DRAM CB can be larger because it's locally_allocated).
+      bool configIsDRAM = false;
+      if (config.outputLayout) {
+        configIsDRAM = !config.outputLayout.hasL1BufferType();
+      }
+      uint64_t cbP = 0;
+      if (it != perOpConfigs.end()) {
+        cbP =
+            configIsDRAM ? it->second.dramCBPeakUsage : it->second.cbPeakUsage;
+      }
```
Comment (bmalesevicTT) [inline test/unittests/Optimizer/L1SpillTestFixture.h:298]: Non-blocking:
dramCBPeakUsage is never set by any test, so the DRAM CB path in makeValidator always returns 0. The logic is correct but the evictForDramCBGrowth scenario — where a demoted op has a larger DRAM CB that affects the cushion math — is unexercised. 
We should add a follow-up test that sets a non-zero dramCBPeakUsage and asserts the expected cushion behaviour

File: test/unittests/Optimizer/L1SpillTestFixture.h [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+  /// Count how many ToMemoryConfigOp ops appear immediately before `op`.
+  size_t reshardsDirectlyBefore(mlir::Operation *op) {
+    size_t n = 0;
+    mlir::Operation *cursor = op->getPrevNode();
+    while (cursor && mlir::isa<ToMemoryConfigOp>(cursor)) {
+      ++n;
+      cursor = cursor->getPrevNode();
+    }
+    return n;
+  }
```
Comment (bmalesevicTT) [inline test/unittests/Optimizer/L1SpillTestFixture.h]: Don't really get the intent behind this func, as it only look back in the IR, doesn't specify which type of reshards is counted. 
1) If used for determining if an op in "nonImplemented" and spills all live tensors, renaming it to numSpillsDirectlyBefore and checking the output of each being DRAM
2) if used for reshards between producers and consumer, it probably needs to check if the reshards are producers of the ops consumers, and not simply next in Schedule. 
3) also, not sure how this would behave with the recently added transientReshards for ops which need L1 input but were affected by spill upstream
Would delete for now and add when needed for a test with clearer intent.

File: test/unittests/Optimizer/TestL1SpillManagement.cpp [hunk 3]
```diff
@@ -0,0 +1,638 @@
...
+//===----------------------------------------------------------------------===//
+// SelfSpillTest
+//===----------------------------------------------------------------------===//
+
+class SelfSpillTest : public L1SpillTestFixture {};
+
+// Op whose output alone exceeds budget cannot be made to fit by evicting
+// anything else. The pass demotes the op's output in place (result type
+// flips from L1 to DRAM) and fires onSelfSpill — this is the Stage-3
+// OOM path in handleOOM (post remove-output-l1-usage-attr).
```
Comment (bmalesevicTT) [inline test/unittests/Optimizer/TestL1SpillManagement.cpp:86]: Nit: would call SelfEvict or SelfDemote or something, since we updated it as it's not really spilling, but that's a refactor in regular L1SpillObserver.h too.

File: test/unittests/Optimizer/TestL1SpillManagement.cpp [hunk 4]
```diff
@@ -0,0 +1,638 @@
...
+
+//===----------------------------------------------------------------------===//
+// NotImplementedPostFlushTest
+//===----------------------------------------------------------------------===//
+
+class NotImplementedPostFlushTest : public L1SpillTestFixture {};
+
+// After evictAllFromL1 resets the tracker, a subsequent allocation must
+// land at a fresh top-of-budget address as if no prior tensors existed.
+// This is the runtime check for Task 1.5's invariant (spillToDramBeforeTrigger
```
Comment (bmalesevicTT) [inline test/unittests/Optimizer/TestL1SpillManagement.cpp]: Task 1.5 reference could be removed lol

File: test/unittests/Optimizer/TestL1SpillManagement.cpp [hunk 5]
```diff
@@ -0,0 +1,638 @@
...
+  finishFunc({opC->getResult(0), usePostFlush->getResult(0)});
+
+  auto [obs] = run();
+
+  // A and B must have been spilled by evictAllFromL1.
+  EXPECT_TRUE(wasSpilled(opA->getResult(0)));
+  EXPECT_TRUE(wasSpilled(opB->getResult(0)));
+  // opPostFlush must NOT be spilled — the reset tracker has 1.3 MiB free.
+  EXPECT_FALSE(wasSpilled(opPostFlush->getResult(0)))
+      << "post-flush allocation should fit on a fully-reset tracker";
```
Comment (bmalesevicTT) [inline test/unittests/Optimizer/TestL1SpillManagement.cpp:191]: Also verify that opPostFlush was not demoted, a corrupted free-list after the tracker reset could cause handleNoFit to call demoteToDram, which wasSpilled would not catch. 
Check either that obs->demotions is empty or that opPostFlush's result type still has an L1 buffer type.
Same for SnapshotReplayManyAllocsTest, and others that do markEvictedAndRebuild

File: test/unittests/Optimizer/TestL1SpillManagement.cpp [hunk 6]
```diff
@@ -0,0 +1,638 @@
...
+  auto *useLarge =
+      addUnary(opLarge->getResult(0), tt, /*l1UsageBytes=*/50 * kKiB);
+  finishFunc({opCB->getResult(0), useLarge->getResult(0)});
+
+  auto [obs] = run();
+
+  EXPECT_TRUE(wasSpilled(opLarge->getResult(0)))
+      << "opLarge should be spilled to clear the CB region";
+  // Observer should record either an eviction or a fragmentation demote.
+  EXPECT_TRUE(!obs->evictions.empty() || !obs->demotions.empty());
```
Comment (bmalesevicTT) [inline test/unittests/Optimizer/TestL1SpillManagement.cpp]: Why would we still check for demotion when we expect eviction here and check it in the line above, demotions should have their own tests

**Review by bmalesevicTT — APPROVED** [review]
Thanks, great to have, some notes to improve clearance further

File: test/unittests/Optimizer/TestL1SpillManagement.cpp (hunk 3 above)
Comment (rpavlovicTT) [reply test/unittests/Optimizer/TestL1SpillManagement.cpp:86]: Kept the name as SelfSpillTest to stay aligned with the existing onSelfSpill observer callback in L1SpillObserver.h. Agree the whole spill/demote naming is muddled — better to rename observer + test together in a separate cleanup so they don't drift. Happy to file a follow-up.
